Mathlib/Algebra/MonoidAlgebra/Degree.lean

```diff
@@ -222,35 +223,306 @@ variable [Semiring R]
 
 section SupDegree
 
-variable [AddZeroClass A] [SemilatticeSup B] [AddZeroClass B] [OrderBot B]
-  (D : A → B)
+variable [AddZeroClass A] [Add B] (D : A → B)
+
+section SemilatticeSup
+
+variable [SemilatticeSup B] [OrderBot B]
 
 /-- Let `R` be a semiring, let `A, B` be two `AddZeroClass`es, let `B` be an `OrderBot`,
 and let `D : A → B` be a "degree" function.
 For an element `f : R[A]`, the `AddMonoidAlgebra R A`, the element `supDegree f : B` is the
 supremum of all the elements in the support of `f`, or `⊥` if `f` is zero.
 Often, the Type `B` is `WithBot A`,
 If, further, `A` has a linear order, then this notion coincides with the usual one,
-using the maximum of the exponents. -/
-@[reducible]
-def supDegree (f : AddMonoidAlgebra R A) : B :=
-  f.support.sup D
+using the maximum of the exponents.
+
+If `A := σ →₀ ℕ` then `R[A] = MvPolynomial σ R`, and if we equip `σ` with a linear order then
+the induced linear order on `Lex A` equips `MvPolynomial` ring with a
+[monomial order](https://en.wikipedia.org/wiki/Monomial_order) (i.e. a linear order on `A`, the
+type of (monic) monomials in `R[A]` that respects addition). We make use of this monomial order
+by taking `D := toLex`, and different monomial orders could be accessed via different type
+synonyms once they are added. -/
+@[reducible] def supDegree (f : AddMonoidAlgebra R A) : B := f.support.sup D
 
-theorem supDegree_add_le (f g : AddMonoidAlgebra R A) :
+/-- If `D` is an injection into a linear order `B`, the leading coefficient of `f : R[A]` is the
+  nonzero coefficient of highest degree according to `D`, or 0 if `f = 0`. In general, it is defined
+  to be the coefficient at an inverse image of `supDegree f`. -/
+noncomputable def leadingCoeff (f : AddMonoidAlgebra R A) : R := f (D.invFun <| f.supDegree D)
```

Comment:
If we only ever take `B := Lex A` or some other type synonym, then we could always take `D` to be an `Equiv`, and then we can use `D.symm` rather than `D.invFun` to define `leadingCoeff`, which would be more convenient at places. However, @adomani [said](https://github.com/leanprover-community/mathlib4/pull/6260#issuecomment-1680236710) he'd like to consider `B := WithBot (Lex A)` to obtain the analogue of `Polynomial.degree` which he thinks is superior to `natDegree`. Personally, I've found `B := Lex A` and `natDegree` convenient to work with, and `WithBot ...` never enjoys the nice cancellative property.   However, I understand a `bot` must be adjoined for something like the Laurent polynomial ring. Fortunately, most results here works for `D` injective but not bijective.

Reply:
I think that it would be safe to wait a little bit longer, before choosing to specialise the definition of degree.  I imagine that "total degree" of a MvPolynomial could benefit from the flexibility.  Also, if we end up working with [elimination ideals](https://pi.math.cornell.edu/~dmehrle/notes/old/alggeo/18Elimination.pdf), I imagine that we will want to use only a subset of the variables to compare monomials.  Or maybe we want to assign arbitrary weights to the variables...  I feel that *most* of the times, `B` is going to be a (type synonym of) `A`, but some-not-so-few times, it may not.

Reply:
Ok, I would probably need to experiment with this.  I can see that it would be easier to always assume that `B` is a type synonym of `A` and `D` is basically the identity.

If, with the current setup, we have to jump through too many hoops, we can always try the following workaround for the cases that I have in mind.  We build the final `MvPolynomial` ring in two steps, first adding all the variables that we do not want to be part of the degree and then all the remaining ones.  E.g. if we want to use as "degree" the degree in `x, y`, but not in `y z`, then we form `R[y, z][x, y]` (I'm mixing formal and informal notation here).  I'm not sure how to work with general weights, but maybe we can worry about this when we use them!

Reply:
According to the [WeightedHomogeneous](https://github.com/leanprover-community/mathlib4/blob/193761250a9869327056b499a7f49d676ed5e3a9/Mathlib/RingTheory/MvPolynomial/WeightedHomogeneous.lean) API, maybe we won't take a type synonym to get the other monomial orders; since all monomial orders can be obtained using [multiple weights](https://en.wikipedia.org/wiki/Monomial_order#Weight_order), a general approach would be taking `B := Fin m → ℕ` (or `ℝ`) and `D` consisting of `m` weighting functions. As long as we have enough weighting functions to break all ties, `D` would be injective and most of the API here apply.

If `D` isn't injective like in the case of elimination ideals you're considering, `supDegree` should still work but we probably shouldn't be using `leadingCoeff` to return a single coefficient (we probably want the "homogeneous component" instead).

Reply:
@alreadydone what do you want me to do with your last comment here?